scanpartitions.8
.TH SCANPARTITIONS "8" "January 2007" "" ""
.SH NAME
scanpartitions \- fstab helper script
.SH SYNOPSIS
\fBscanpartitions\fR [\-l|\-\-labels] [\-u|\-\-uuids] [devices]
.SH DESCRIPTION
Parses /proc/partitions and generates a device name, mount point
and filesystem type suitable for use in /etc/fstab for each valid
partition. vol_id is used to gather information about the volume
label and filesystem uuid/type for each partition.
.PP
Using UUID or LABEL disk identification in /etc/fstab is more robust
than using the device name; adding or removing a system disk changes
disk device names but not the filesystem uuid or volume label.
.PP
LABEL is given priority over UUID when the \fB\-\-labels\fR and 
\fB\-\-uuids\fR options are used together.
.PP
Linux non-native filesystems (eg. ntfs, vfat, hfsplus) may not be
able to make use of the LABEL or UUID volume identifiers. These
filesystems can use the corresponding /dev/disk/by-label/ or
/dev/disk/by-uuid/ device nodes when existing, otherwise fall back
to the base device node (eg /dev/hda1) and not benefit from the
advantages described above.
.PP
.SH OPTIONS
.TP
\fB\-l|\-\-labels\fR
Enable the use of the volume LABEL identifier.
.TP
\fB\-u|\-\-uuids\fR
Enable the use of the filesystem UUID identifier.
.PP
.SH USAGE
\fBscanpartitions\fR will parse \fI/proc/partitions\fR if no block
device names are given as arguments.
.PP
The following example will print a line for all supported volumes
contained in \fI/proc/partitions\fR, using UUID metainfo for volume
name:
.PP
.RS
\fBscanpartitions \-\-uuids\fR
.RE
.PP
The following example with print a line for block devices /dev/sda1
and /dev/sda2, using LABEL metainfo where possible:
.PP
.RS
\fBscanpartitions \-\-labels /dev/sda1 /dev/sda2\fR
.RE
.PP
.SH AUTHOR
This manual page was written by Kel Modderman <kel@otaku42.de> for
the Debian GNU system (but may be used by others).


debian/changelog
scanpartitions (0.7.5) UNRELEASED; urgency=low

  * NOT RELEASED YET

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sun, 14 Jan 2007 15:29:40 +0100

scanpartitions (0.7.4) unstable; urgency=low

  [ Michiel de Boer ]
  * Misc improvements
    - use real argument handling
    - parse vol_id and assign values to assoc. arrays
    - reorder functions in order of use
    - create parse_all function for the case when no devices are given as
      arguments (most common usage case).

  [ Kel Modderman ]
  * Tiny adjustment to improve formatting of manpage synopsis section.
  * Revert to using split() for parsing vol_id output instead of globally
    altering IFS.
  * Use more descriptive var. name for disk array in BEGIN block (p -> disk).
  * Use consistent coding style w.r.t. to if blocks. I know its not "pretty"
    but I'd rather have similar style used without script.
  * Clarify some comments about the relevance of block numbers in
    /proc/partitions w.r.t. extended partitions.

 -- Kel Modderman <kel@otaku42.de>  Fri, 12 Jan 2007 07:55:05 +1000

scanpartitions (0.7.3) unstable; urgency=low

  [ Kel Modderman ]
  * Use explicit install commands in debian/rules instead of debian/install.
  * Replace scanpartitions shell script/fstype gawk script combo for a single
    awk script that uses vol_id to determine partition properties.
  * Drop dependency on gawk and disktype.
  * Add dependency on udev, which provides /lib/udev/vol_id.
  * Add scanpartitions.8 manpage.
  * Add myself to uploaders.
  * Update debian/copyright accordingly.

  [ Stefan Lippers-Hollmann ]
  * remove fstype and scanpartitions from source.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sun,  7 Jan 2007 13:40:49 +0100

scanpartitions (0.7.2) unstable; urgency=low

  [ Joaquim Boura ]
  * fix lintian errors.
  * Initial version for sidux

  [ Stefan Lippers-Hollmann ]
  * tweak debian/control.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sun,  3 Dec 2006 02:26:50 +0100

scanpartitions-kanotix (0.7.1) unstable; urgency=low

  * remove cdbs auto-update feature.
  * remove knoppix virtual packages.
  * bump compat level to 5.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sat, 17 Dec 2005 03:24:19 +0100

scanpartitions-kanotix (0.7) unstable; urgency=low

  * change /mnt to /media

 -- Joerg Schirottke <master@kanotix.com>  Thu,  3 Nov 2005 12:37:14 +0100

scanpartitions-kanotix (0.6.1) unstable; urgency=low

  * fix FSF address.
  * add /dev/cciss/c0d0 support.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Fri, 21 Oct 2005 02:18:29 +0200

scanpartitions-kanotix (0.6) unstable; urgency=low

  * re- indent.
  * switch to locsmif's re- write of fstype based on disktype.
  * we don't depend on file any longer.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Mon, 10 Oct 2005 14:06:23 +0200

scanpartitions-kanotix (0.5.15) unstable; urgency=low

  * packaging overhaul.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Tue, 23 Aug 2005 18:37:10 +0200

scanpartitions-kanotix (0.5.14) unstable; urgency=low

  * import hwraid support by kelmo. 

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sat, 30 Apr 2005 14:47:21 +0200

scanpartitions-kanotix (0.5.13) unstable; urgency=low

  * important updates from Knoppix 3.8
    - improve swap detection.
    - handle /dev/ub?.
  * conflict with scanpartitions-knoppix.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Tue, 22 Mar 2005 17:54:52 +0100

scanpartitions-kanotix (0.5-12) unstable; urgency=low

  * fix dependencies. 

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Thu, 23 Dec 2004 01:33:34 +0100

scanpartitions-kanotix (0.5-11) unstable; urgency=low

  * Fork for Kanotix compatibility.
  * add reiser4 detection (note: parted can't detect reiser4 yet).

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Tue, 21 Dec 2004 22:10:49 +0100

scanpartitions-knoppix (0.5-10) unstable; urgency=low

  * Take into account /dev/ataraid/d?p*

 -- Klaus Knopper <knopper@linuxtag.net>  Mon, 16 Aug 2004 07:41:14 +0200

scanpartitions-knoppix (0.5-9) unstable; urgency=low

  * Don't add partitions larger than disksize to fstab (workaround for broken USB storage devices)

 -- Klaus Knopper <knopper@linuxtag.net>  Sun,  8 Aug 2004 16:28:42 +0200

scanpartitions-knoppix (0.5-8) unstable; urgency=low

  * Make fstype a separate script

 -- Klaus Knopper <knopper@linuxtag.net>  Thu, 22 Apr 2004 02:14:31 +0200

scanpartitions-knoppix (0.5-7) unstable; urgency=low

  * Shift sort order of entries, because bootable partitions seem to have an
    *x86* entry, even if they are not (V)FAT.

 -- Klaus Knopper <knopper@linuxtag.net>  Mon, 19 Apr 2004 20:39:03 +0200

scanpartitions-knoppix (0.5-6) unstable; urgency=low

  * No test mounts (hang with USB memorysticks sometimes)

 -- Klaus Knopper <knopper@linuxtag.net>  Sat,  3 Apr 2004 21:36:52 +0200

scanpartitions-knoppix (0.5-5) unstable; urgency=low

  * Exclude ram* and cloop* from /proc/partitions (Kernel 2.6 bug/feature?)

 -- Klaus Knopper <knopper@linuxtag.net>  Mon,  1 Mar 2004 09:39:53 +0100

scanpartitions-knoppix (0.5-4) unstable; urgency=low

  * Replaced testmounts with echos for "well-known" filesystems for faster scanning

 -- Klaus Knopper <knopper@linuxtag.net>  Tue,  3 Dec 2002 21:57:53 +0100

scanpartitions-knoppix (0.5-3) unstable; urgency=low

  * Glue with knoppix-rebuildfstab

 -- Klaus Knopper <knopper@linuxtag.net>  Tue, 26 Nov 2002 23:04:49 +0100

scanpartitions-knoppix (0.5-2) unstable; urgency=low

  * Workaround a /proc/partitions read kernel bug

 -- Klaus Knopper <knopper@linuxtag.net>  Sat,  5 Oct 2002 21:39:47 +0200

scanpartitions-knoppix (0.5-1) unstable; urgency=low

  * Initial Release.

 -- Klaus Knopper <knopper@linuxtag.net>  Mon, 12 Aug 2002 13:56:53 +0200



scanpartitions.awk
#!/usr/bin/awk -f
#
# Copyright (C) 2007 Kel Modderman <kel@otaku42.de>
#		     Michiel de Boer <locsmif@kanotix.com>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this package; if not, write to the Free Software 
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, 
# MA 02110-1301, USA.
#
# On Debian GNU/Linux systems, the text of the GPL license can be
# found in /usr/share/common-licenses/GPL.

function blockdev_exists(node,	retval)
{
	cmd = "test -b \"" node  "\""
	(system(cmd) == 0) ? retval = 0 : retval = 1
	return retval
}

function parse_vol_id(name)
{
	# ensure these var's are empty
	delete id
	delete vol_id
	mntdev = mntpnt = ""

	# ensure name has /dev/ stripped from it
	sub(/^\/dev\//, "", name)
	
	# ensure `dev' is prefixed with /dev/
	dev = "/dev/" name

	# test block device node for existance
	if (blockdev_exists(dev) != 0)
		return 0
		
	# device name blacklist
	if (name ~ /^(ram|cloop|loop).+/)
		return 0

	# run vol_id on dev, discard stderr
	cmd = "/lib/udev/vol_id " dev " 2>/dev/null"
	# form assoc. array `id'
	while (cmd | getline) {
		# indices: label label_safe type usage uuid version
		split($0, vol_id, "=")
		sub(/^ID_FS_/, "", vol_id[1])
		id[tolower(vol_id[1])] = vol_id[2]
	}

	# return if we cannot handle fs usage type
	if (id["usage"] !~ /(filesystem|other)/)
		return 0

	# return if we can't reliably use the partitions filesystem
	if (!id["type"] || id["type"] == "unknown")
		return 0

	# use metainfo for linux native filesystems
	if (id["type"] ~ /^(ext[234]|jfs|reiser|swap|xfs)/) {
		if (labels && id["label"])
			mntdev = "LABEL=" id["label"]
		else if (uuids && id["uuid"])
			mntdev = "UUID=" id["uuid"]
	}
	# workaround for non-perfect non-native fs support
	else {
		if (labels && id["label"])
			mntdev = "/dev/disk/by-label/" id["label"]
		else if (uuids && id["uuid"])
			mntdev = "/dev/disk/by-uuid/" id["uuid"]
		
		# block device may not yet exist if vfat/ntfs volume was just created
		if (blockdev_exists(mntdev) != 0)
			mntdev = dev
	}
	
	# fall back to raw device name if uuid/label not found or wanted
	if (!mntdev)
		mntdev = dev

	# no mount point for swap
	mntpnt = (id["type"] == "swap") ? "none" : "/media/" name
	
	# print something useful for a fstab helper program
	print(mntdev, mntpnt, id["type"])
}

function parse_all() {
	# parse /proc/partitions
	while (getline < "/proc/partitions") {
		# major minor #blocks name
		#
		# Skip if illegal entry, or if first column is not a number, or if
		# number of blocks indicates ext. partition
		# Note: $2 % 16 could be used to determine major block device,
		# such as hda. However superfloppies are possible, so leaving that
		# out.
		if (! $4 || $1 !~ /[0-9]+/ || $3 < 2)
			continue
		else		
			parse_vol_id($4)
	}
}

function usage() {
	print("Usage: scanpartitions [-l|--labels] [-u|--uuids] [device]")
	exit(0)
}

BEGIN {
	if (ARGC > 1) {
		# test devices given as arguments
		for (i = 1; i < ARGC; i++) {
			if (ARGV[i] ~ /^(-u|--uuids)/)
				uuids=1
			else if (ARGV[i] ~ /^(-l|--labels)/)
				labels=1
			else if (ARGV[i] ~ /^-/)
				usage()
			else
				disk[++j] = ARGV[i]
		}
	}
	
	if (disk[1])
		for (i in disk) parse_vol_id(disk[i])
	else
		parse_all()
}

